[FLINK-28480] Forward timeControllerExecution time as histogram for JOSDK Metrics interface

SteNicholas opened the pull request:
## What is the purpose of the change

Currently the JOSDK metrics forwarder logic doesn't implement the `timeControllerExecution` function.  We should implement it with the following logic.

1. Measure execution time for successful failed executions
2. Based on the name of the ControllerExectution (reconcile/cleanup) and controller name, track the following histogram metrics metrics

`JOSDK.{ControllerExecution.controllerName}.{ControllerExecution.name}.{ControllerExecution.successTypeName}/failed`

## Brief change log

  - Introduces the `Histogram` metric in the `timeControllerExecution` with the metric name format `JOSDK.{ControllerExecution.controllerName}.{ControllerExecution.name}.{ControllerExecution.successTypeName}/failed`.

## Verifying this change

  - *`OperatorJosdkMetricsTest` adds the `testTimeControllerExecution` to verify whether the `Histogram` metric is correct.*
  - The `Histogram` metric in the `timeControllerExecution` is as follows:
  
```
-- Histograms -----------------------------------------------------------------
flink-kubernetes-operator-7679966cd6-d4p2g.k8soperator.default.flink-kubernetes-operator.system.JOSDK.flinkdeploymentcontroller.reconcile.resource.Nanos: count=6, min=6517100, max=1198002100, mean=3.375813833333334E8, stddev=4.5900281423332655E8, p50=1.997605E8, p75=6.04624825E8, p95=1.1980021E9, p98=1.1980021E9, p99=1.1980021E9, p999=1.1980021E9
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: (yes / **no**)
  - Core observer or reconciler logic that is regularly executed: (yes / **no**)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

SteNicholas commented:
cc @gyfora @morhidi , PTAL.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/metrics/OperatorJosdkMetrics.java:
```diff
@@ -43,21 +47,41 @@ public class OperatorJosdkMetrics implements Metrics {
+        try {
+            T result = execution.execute();
+            String successType = execution.successTypeName(result);
+            histogram(execution.controllerName(), execution.name(), successType)
```
Instead if using the controllerName directly could we please map this to `FlinkDeployment` / `FlinkSessionJob` based on the name? This would make it consistent with other metrics

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/metrics/OperatorJosdkMetrics.java:
```diff
@@ -113,6 +137,19 @@ public void failedReconciliation(ResourceID resourceID, Exception exception) {
+        return histograms.computeIfAbsent(
+                String.join(".", group.getScopeComponents()),
+                s ->
+                        finalGroup.histogram(
```
Instead of `Nanos` I suggest we use Seconds and name it to `TimeSeconds` to make it consistent with other metrics.

SteNicholas replied:
@gyfora, IMO, the `TimeNanos` makes sense because this use the `clock.relativeTimeNanos` not the seconds. WDYT?

gyfora replied:
Please convert the nano time to seconds before calling it `TimeSeconds` :D

gyfora reviewed (changes requested):
Thanks @SteNicholas , looks good, added some minor comments

SteNicholas commented:
@gyfora, I have addressed above comments and verified locally. The metric is as follows:
```
-- Histograms -----------------------------------------------------------------
flink-kubernetes-operator-7679966cd6-v4qlw.k8soperator.default.flink-kubernetes-operator.system.JOSDK.FlinkDeployment.reconcile.resource.TimeSeconds: count=6, min=0, max=0, mean=0.0, stddev=0.0, p50=0.0, p75=0.0, p95=0.0, p98=0.0, p99=0.0, p999=0.0
```

gyfora reviewed (approved):
Thanks @SteNicholas looks good! Merging this after CI
